### PR TITLE
Lint everything

### PR DESCRIPTION
**Pull request by stefanv**
See https://github.com/scipy/scipy/pull/17812

### CODE REVIEW

**Comment by stefanv** [comment]
For merging after #17812.

/cc @tupui @mattip @mdhaber

**Comment by tupui** [comment]
Everyone, let's try to not break this so please refrain from merging.

If you were not there during the last community meeting, we gave the green light for this change. Thanks a ton @stefanv for doing these tedious manual changes.

Extra cc of current people who merge (hope I don't forget anyone, if I do sorry)
@tylerjereddy @rgommers @WarrenWeckesser @andyfaff @ev-br

File: scipy/optimize/_linprog_highs.py [hunk 1]
```diff
@@ -19,7 +19,6 @@
 from warnings import warn
 from ._highs._highs_wrapper import _highs_wrapper
 from ._highs._highs_constants import (
-    CONST_I_INF,
```
Comment (tupui) [inline scipy/optimize/_linprog_highs.py:22]: Not sure

File: scipy/optimize/_linprog_ip.py [hunk 2]
```diff
@@ -28,13 +28,11 @@
...
-    import sksparse
-    from sksparse.cholmod import cholesky as cholmod
     from sksparse.cholmod import analyze as cholmod_analyze
 except ImportError:
     has_cholmod = False
 try:
-    import scikits.umfpack  # test whether to use factorized
+    pass  # test whether to use factorized
 except ImportError:
     has_umfpack = False
```
Comment (tupui) [inline scipy/optimize/_linprog_ip.py:37]: Not sure

File: scipy/special/_orthogonal.pyi [hunk 3]
```diff
@@ -1,7 +1,6 @@
 from typing import (
     Any,
     Callable,
-    List,
     Literal,
     Optional,
```
Comment (tupui) [inline scipy/special/_orthogonal.pyi:5]: Should be gone no?

File: scipy/optimize/_linprog_ip.py [hunk 4]
```diff
@@ -28,13 +28,11 @@
...
 has_cholmod = True
 try:
-    import sksparse
-    from sksparse.cholmod import cholesky as cholmod
     from sksparse.cholmod import analyze as cholmod_analyze
 except ImportError:
     has_cholmod = False
 try:
-    import scikits.umfpack  # test whether to use factorized
+    pass  # test whether to use factorized
```
Comment (mdhaber) [inline scipy/optimize/_linprog_ip.py:35]: This appears to change something. I don't want to dig into whether it matters or not. These solvers are deprecated, so can we just `# noqa`?

File: scipy/setup.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 def configuration(parent_package='',top_path=None):
     from numpy.distutils.system_info import get_info
-    lapack_opt = get_info("lapack_opt")
```
Comment (tupui) [inline scipy/setup.py:3]: Not sure

File: scipy/optimize/tests/test_linprog.py [hunk 6]
```diff
@@ -23,8 +23,7 @@
 
 has_cholmod = True
 try:
-    import sksparse
-    from sksparse.cholmod import cholesky as cholmod
+    pass
 except ImportError:
     has_cholmod = False
```
Comment (tupui) [inline scipy/optimize/tests/test_linprog.py:28]: Not sure

File: scipy/optimize/_trustregion_constr/tests/test_projections.py [hunk 7]
```diff
@@ -7,7 +7,6 @@
                            assert_equal, assert_allclose)
 
 try:
-    from sksparse.cholmod import cholesky_AAt
     sksparse_available = True
     available_sparse_methods = ("NormalEquation", "AugmentedSystem")
 except ImportError:
```
Comment (tupui) [inline scipy/optimize/_trustregion_constr/tests/test_projections.py:12]: Not sure

File: scipy/signal/_ltisys.py [hunk 8]
```diff
@@ -51,7 +51,7 @@ def __new__(cls, *system, **kwargs):
             raise NotImplementedError('The LinearTimeInvariant class is not '
                                       'meant to be used directly, use `lti` '
                                       'or `dlti` instead.')
-        return super(LinearTimeInvariant, cls).__new__(cls)
+        return super().__new__(cls)
```
Comment (tupui) [inline scipy/signal/_ltisys.py:54]: Not sure on this file

**Review by tupui — COMMENTED** [review]
Thanks again! I went trough all the files... Here are my notes. My time is up for this WE. I will have a look on Monday if this is still up by then.

File: scipy/optimize/tests/test_linprog.py [hunk 9]
```diff
@@ -63,7 +62,7 @@ def _assert_success(res, desired_fun=None, desired_x=None,
     # desired_fun: desired objective function value or None
     # desired_x: desired solution or None
     if not res.success:
-        msg = "linprog status {0}, message: {1}".format(res.status,
+        msg = "linprog status {}, message: {}".format(res.status,
                                                         res.message)
```
Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:66]:
```suggestion
        msg = "linprog status {}, message: {}".format(res.status,
                                                      res.message)
```

**Review by mdhaber — COMMENTED** [review]
I took a quick look at `optimize` and `stats`. I suppose if tests pass, I don't object.

File: scipy/optimize/tests/test_linprog.py (hunk 6 above)
Comment (mdhaber) [reply scipy/optimize/tests/test_linprog.py:28]: Yeah I'd just `noqa` all of the imports in `linprog` files.

**Comment by stefanv** [comment]
Everyone who wants to, please feel free to push commits to this PR! I won't be able to spend much time on it the rest of the weekend, but I also don't want to block things.

**Review by tylerjereddy — COMMENTED** [review]
Some tests error out because of circular import scenarios that were "corrected" (broken) by moving all imports to top of module. For example, one category of testing errors is fixed by this diff (I'll see if I can paste diffs for a few other fixes..). One could possibly refactor to avoid the circular imports, though that may be more work.

```diff
--- a/scipy/linalg/_matfuncs_sqrtm.py
+++ b/scipy/linalg/_matfuncs_sqrtm.py
@@ -15,13 +15,15 @@ from ._misc import norm
 from .lapack import ztrsyl, dtrsyl
 from ._decomp_schur import schur, rsf2csf
 
-from ._matfuncs_sqrtm_triu import within_block_loop
 
 
 class SqrtmError(np.linalg.LinAlgError):
     pass
 
 
+from ._matfuncs_sqrtm_triu import within_block_loop
+
+
 def _sqrtm_triu(T, blocksize=64):
     """
     Matrix square root of an upper triangular matrix.

```

**Comment by stefanv** [comment]
Thanks, Tyler, feel free to apply. We can also add more files to the ignores in ruff.toml

**Comment by tylerjereddy** [comment]
Ok, I pushed in a commit that allows the full suite to pass for me locally. I didn't touch `ruff.toml` exceptions, but that shouldn't be too hard to amend later. I did try to use `git checkout -p` and apply sub-patches to reduce reverting anything that didn't need to be reverted, though probably still is a superset.

**Comment by stefanv** [comment]
Thanks, Tyler!

**Comment by ev-br** [comment]
Spot-checked a few files in interpolate and elsewhere, and this seems mostly harmless. So I'm going to keep being -0 on these, and will stay out of discussions.

**Comment by ilayn** [comment]
Same with @ev-br I think the best thing is to be done with it, looks pretty OK except the multiline handling with format but f-strings inline strings. But I don't think it is a bad choice per se. 

Already caused unnecessary work with ruff suggestion 😄 apologies for that. So I'll try to get out of the way too.

File: scipy/_lib/uarray.py [hunk 10]
```diff
@@ -23,9 +23,7 @@
 
 if _has_uarray:
     from uarray import *
-    from uarray import _Function
 else:
     from ._uarray import *
-    from ._uarray import _Function
```
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/_lib/uarray.py:29]: Ok this is needed at least by the doc build. See the CI failure https://app.circleci.com/pipelines/github/scipy/scipy/18026/workflows/304aeadd-108e-4b49-9eb3-ae1c5d20913e/jobs/61668

**Comment by stefanv** [comment]
Thanks, all; please note that this PR will not pass until we merge #17812.

We need to move those two in tandem, and do it fairly quickly, otherwise we'll get out of sync with `main`.  I'm happy to keep making adjustments, but I don't want to do the same work over and over, if that makes sense.

My suggestion is that we merge #17812 now, spend a few hours shoring up this PR, and then get that in too.

**Comment by tupui** [comment]
Agreed, I merged. Thanks for all the work you are putting in this @stefanv.

**Comment by stefanv** [comment]
Crossing fingers that this catches the remaining linting errors.

File: scipy/special/tests/test_wrightomega.py [hunk 11]
```diff
@@ -52,7 +52,7 @@ def test_wrightomega_singular():
     for p in pts:
         res = sc.wrightomega(p)
         assert_equal(res, -1.0)
-        assert_(np.signbit(res.imag) == False)
+        assert_(np.signbit(res.imag) is False)
```
Comment (WarrenWeckesser) [inline scipy/special/tests/test_wrightomega.py:55]: We can't use `is` here.  The return value of `np.signbit(res.imag)` is an instance of `np.bool_`.  It is equal to `False`, but it is a different object.  How about using
```
        assert_(not np.signbit(res.imag))
```
instead?  (Or just go back to `==`.)

**Comment by tylerjereddy** [comment]
> We can't use is here.

I thought I corrected that in my partial-reversion commit 94d6e70, maybe it got reverted from the reversion again??

**Comment by tylerjereddy** [comment]
Hmm, yeah, looking at the most recent commit some of those changes revert fixes I thought were needed.

**Comment by stefanv** [comment]
I think one fix was because I replaced it with bad logic (the isinstance). The other one with the bool comparison I need to fix.

File: scipy/_lib/tests/test__gcutils.py [hunk 12]
```diff
@@ -75,7 +75,7 @@ class C:
         # to function correctly.  It is assigned to c, but c itself is
         # not referenced in the body of the with, it is only there for
         # the refcount.
-        with assert_deallocated(C) as c:
+        with assert_deallocated(C):
```
Comment (WarrenWeckesser) [inline scipy/_lib/tests/test__gcutils.py:78]: This change causes the failure
```
================================================================== FAILURES ==================================================================
_______________________________________________________ test_assert_deallocated_nodel ________________________________________________________
[gw3] linux -- Python 3.10.8 /home/warren/py3.10.8/bin/python3
scipy/_lib/tests/test__gcutils.py:72: in test_assert_deallocated_nodel
    with pytest.raises(ReferenceError):
E   Failed: DID NOT RAISE <class 'scipy._lib._gcutils.ReferenceError'>
        C          = <class 'scipy._lib.tests.test__gcutils.test_assert_deallocated_nodel.<locals>.C'>
========================================================== short test summary info ===========================================================
FAILED scipy/_lib/tests/test__gcutils.py::test_assert_deallocated_nodel - Failed: DID NOT RAISE <class 'scipy._lib._gcutils.ReferenceError'>
============================================= 1 failed, 86 passed, 2 skipped, 1 xfailed in 8.03s =============================================

```
It looks like the test requires the reference to be held by `c`.

A possible fix that makes `ruff` happy is
```
        with assert_deallocated(C) as _:
```

File: scipy/stats/_kde.py [hunk 13]
```diff
@@ -260,7 +260,7 @@ def evaluate(self, points):
                 points = reshape(points, (self.d, 1))
                 m = 1
             else:
-                msg = "points have dimension %s, dataset has dimension %s" % (d,
+                msg = "points have dimension {}, dataset has dimension {}".format(d,
                     self.d)
```
Comment (WarrenWeckesser) [inline scipy/stats/_kde.py:264]: Was there a reason for not converting this to an f-string, like most of the other formatted string changes?

File: scipy/stats/_stats_mstats_common.py [hunk 14]
```diff
@@ -318,8 +318,8 @@ def theilslopes(y, x=None, alpha=0.95, method='separate'):
 
     """
     if method not in ['joint', 'separate']:
-        raise ValueError(("method must be either 'joint' or 'separate'."
-                          "'{}' is invalid.".format(method)))
+        raise ValueError("method must be either 'joint' or 'separate'."
+                          "'{}' is invalid.".format(method))
```
[the remark this answers is not in the document]
Comment (WarrenWeckesser) [reply scipy/stats/_stats_mstats_common.py:322]: It should be, and running `flake8 --isolated` on this file reports

```
_stats_mstats_common.py:322:27: E127 continuation line over-indented for visual indent
```
I just tried running `ruff check -n --isolated` on this file, but it did not report an issue.   I'm just getting started with `ruff`, so maybe there is an obvious reason for that.

I noticed one or two other small tweaks that should be made, but that can be done in a follow-up, as long as folks are OK with a few more style-related changes after this PR.

[the remark this answers is not in the document]
Comment (stefanv) [reply scipy/stats/_stats_mstats_common.py:322]: > It should be, and running `flake8 --isolated` on this file reports
> 
> ```
> _stats_mstats_common.py:322:27: E127 continuation line over-indented for visual indent
> ```
> 
> I just tried running `ruff check -n --isolated` on this file, but it did not report an issue. I'm just getting started with `ruff`, so maybe there is an obvious reason for that. According to the [`ruff` documentation](https://github.com/charliermarsh/ruff#pycodestyle-e-w), `ruff` doesn't support `E127`. I guess this falls in the category of "stylistic rules made obsolete by the use of an autoformatter, like [Black](https://github.com/psf/black)."
> 
> I noticed one or two other small tweaks that should be made, but that can be done in a follow-up, as long as folks are OK with a few more style-related changes after this PR.

Ruff does not implement a few of the stylistic rules, as it is mostly used in conjunction with black. See the discussion on the related lining tool PR. But I believe ruff is open to adding these for projects that don't use black.

File: scipy/stats/_kde.py (hunk 13 above)
Comment (stefanv) [reply scipy/stats/_kde.py:264]: Just that those changes were made by ruff's equivalent of pyupgrade, I didn't touch much by hand. Happy to upgrade them.

File: scipy/_lib/tests/test__gcutils.py (hunk 12 above)
Comment (stefanv) [reply scipy/_lib/tests/test__gcutils.py:78]: Thanks!

File: scipy/stats/_kde.py (hunk 13 above)
Comment (WarrenWeckesser) [reply scipy/stats/_kde.py:264]: > Happy to upgrade them.

Not necessary.  I see that it happened elsewhere too.  That kind of refinement can be done in a follow-up, if there is agreement on continuing these types of stylistic improvement.

**Comment by tupui** [comment]
Looks like CI is back green. Any more comments or suggestions before we try this out?

**Comment by stefanv** [comment]
> As the `dev.py lint` now checks cython files will these need fixing up too?

The pxi and pxd checking only got added to cython-lint yesterday. When that is enabled, we'd need to make a bunch of pxi and pxd fixes. Otherwise the pyx files are fine, I think.

**Comment by tupui** [comment]
Alright, I am merging so we can move forward. Thanks all!

**Comment by stefanv** [comment]
Thanks, all. I know there will likely be tune-ups necessary. Feel free to open issues and assign me to them.

File: .git-blame-ignore-revs [hunk 15]
```diff
@@ -0,0 +1,2 @@
+# Adopt ruff linter
+70370adaec7c75cb69b5e2a41bc5a6aa1d6b273f
```
Comment (WarrenWeckesser) [inline .git-blame-ignore-revs:2]: This SHA doesn't appear to exist. Does it need to be updated?

Comment (tupui) [reply .git-blame-ignore-revs:2]: Yes this is because I squashed so we can have a single commit/sha

Comment (WarrenWeckesser) [reply .git-blame-ignore-revs:2]: OK, see https://github.com/scipy/scipy/pull/17897.

**Comment by j-bowhay** [comment]
> > As the `dev.py lint` now checks cython files will these need fixing up too?
> 
> The pxi and pxd checking only got added to cython-lint yesterday. When that is enabled, we'd need to make a bunch of further fixes. Otherwise the pyx files are fine, I think.

I think there might be issues in the pyx file see https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=22982&view=logs&j=fdd89725-f657-5341-7d77-3c5c44c4f5b6&t=325fc03e-e0b9-5339-adaf-099a49a14600
I think it would be nice to tidy these up too otherwise it makes the reviewing quite difficult

**Comment by stefanv** [comment]
> I think there might be issues in the pyx file see https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=22982&view=logs&j=fdd89725-f657-5341-7d77-3c5c44c4f5b6&t=325fc03e-e0b9-5339-adaf-099a49a14600
> I think it would be nice to tidy these up too otherwise it makes the reviewing quite difficult

Yeah, I see also that the pycodestyle checking is quite invasive, so perhaps I'll switch that off and correct the other errors.

**Comment by h-vetinari** [comment]
Congrats for this change!

> Thanks, all. I know there will likely be tune-ups necessary. Feel free to open issues and assign me to them.

From @FranzForstmayr in #15929:

> ruff won't let me commit the `doc/source/conf.py` file, as the special variable tags is not defined and ignoring the error for this line does not work. See [charliermarsh/ruff#2561](https://github.com/charliermarsh/ruff/issues/2561)

not sure if it's worth an issue, wanted to comment here first

**Comment by FranzForstmayr** [comment]
> not sure if it's worth an issue, wanted to comment here first

This was my fault :)

**Pull request merged**